src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ReuploadFailedEnvelopeTask.java

```diff
@@ -0,0 +1,51 @@
+package uk.gov.hmcts.reform.bulkscanprocessor.tasks;
+
+import net.javacrumbs.shedlock.core.SchedulerLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.bulkscanprocessor.entity.Envelope;
+import uk.gov.hmcts.reform.bulkscanprocessor.tasks.processor.EnvelopeProcessor;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * This class is a task executed by Scheduler as per configured interval.
+ * It will retrieve failed to upload envelopes from DB and try to re-upload.
+ * <ol>
+ *     <li>Get batch of failed to upload envelopes</li>
+ *     <li>Retrieve relevant zip file from Blob Storage</li>
+ *     <li>Try to upload pdfs</li>
+ * </ol>
+ */
+@Component
+@ConditionalOnProperty(value = "scheduling.enabled", matchIfMissing = true)
+public class ReuploadFailedEnvelopeTask {
```

Comment:
Currently this ReuploadFailedEnvelopeTask fetches first 20 records at a time. It should fetch all configured jurisdiction and then process failed records for each jurisdiction in parallel for e.g we should process sscs and divorce failed records separately and parallely. Within each jurisdiction you can batch them i.e process few records at a time but that should be configurable and not hardcoded.

```
ReuploadFailedEnvelopeTask.java

private EnvelopeAccessProperties accessProps;

    ExecutorService executor = Executors.newFixedThreadPool(
            Runtime.getRuntime().availableProcessors()+1);
    
accessProps.getMappings().forEach(
    mapping -> {
 			executor.execute(
              new FailedDocUploadProcessor().process(mapping.getJurisdiction())
             );
    })

```

```
FailedDocUploadProcessor.java 
    1. Fetch envelopes for the given jurisdiction with status upload failure.You can fetch few records at a time or fetch all(we don't expect many failures unless doc upload service is down)
    2. Try uploading the pdf's
    3. Update scannable items 
```

Reply:
or you could also do this I believe

`accessProps.getMappings()
   .parallelStream()
   .forEach(m -> new FailedDocUploadProcessor().process(m.getJurisdiction()))`

With the caveat that it uses the fork/join thread pool which is shared. On the other hand I believe, the explicitly created thread pool above needs to be also explicitly shutdown.

Reply:
You have to be very careful before using parallel stream. The problem is that all parallel streams use common fork-join thread pool and if you submit a long-running task, you effectively block all threads in the pool.

Yes executor needs to be shutdown explicitly.